x-pack/plugin/core/src/main/java/org/elasticsearch/license/LicenseService.java

```diff
@@ -7,67 +7,79 @@
 
 package org.elasticsearch.license;
 
-import org.elasticsearch.common.settings.Setting;
-import org.elasticsearch.core.TimeValue;
-import org.elasticsearch.protocol.xpack.XPackInfoResponse;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.support.master.AcknowledgedResponse;
+import org.elasticsearch.cluster.metadata.Metadata;
+import org.elasticsearch.common.component.LifecycleComponent;
+import org.elasticsearch.protocol.xpack.license.PutLicenseResponse;
 
-import java.util.List;
-import java.util.Set;
-import java.util.stream.Collectors;
-import java.util.stream.Stream;
+/**
+ * Interface to read the current license. Consumers should generally not need to read the license directly and should instead
+ * prefer {@link XPackLicenseState} to check if a feature is allowed by the license. This interface is not intended to be implemented
+ * by alternative implementations and exists for internal use only.
+ */
+public interface LicenseService extends LifecycleComponent {
```

Comment:
Added the contract and javadoc in this PR to help separate the discussion of the contract vs. implementation. However, in practice this contract is really just derived from the implementation with the following goals :
* Attempt to keep the abstraction as leak free as possible .... the Metadata in the contract is a bit of a leak but kinda unavoidable
* Attempt to seperate the mutable vs. non-mutable interfaces. We inject these implementations and almost always only need the immutable view.

Reply:
decided to defer introduction of contract so this has been removed.